File: src/C/pfcholesky.c
#include <math.h>
#include "chompack.h"

void ddrsv(const int *n, const char *trans, double * restrict v, double * restrict l, double * restrict b, double * restrict x) {
  /*
    Solves a system of the form L*y = x or L.T*y = x where L is the
    Cholesky factor of a matrix of the form I + a*v*v.T. On exit, x
    contains the solution y.
   */
  int k;
  double s=0;
  if (*trans=='N') {
    for (k=0;k<*n;k++) {
      x[k] = (x[k]-v[k]*s)/l[k];
      s += x[k]*b[k];
    }
  }
  else if (*trans=='T') {
    for (k=*n-1;k>=0;k--) {
      x[k] = (x[k]-b[k]*s)/l[k];
      s += x[k]*v[k];
    }
  }
  return;
}

void ddrmv(const int *n, const char *trans, double * restrict v, double * restrict l, double * restrict b, double * restrict x) {
  /*
    Computes the matrix-vector product x := L*x or x := L.T*x where L
    is the Cholesky factor of a matrix of the form I + a*v*v.T.
   */
  int k;
  double tmp,s=0;
  if (*trans=='N') {
    for (k=0;k<*n;k++) {
      tmp = x[k]*b[k];
      x[k] = l[k]*x[k]+v[k]*s;
      s += tmp;
    }
  }
  else if (*trans=='T') {
    for (k=*n-1;k>=0;k--) {
      tmp = x[k]*v[k];
      x[k] = l[k]*x[k]+b[k]*s;
      s += tmp;
    }
  }
  return;
}


int dpftrf(const int *n, const int *k, double * restrict a, double * restrict V, const int *ldv, double * restrict L, const int *ldl, double * restrict B, const int *ldb) {
  /*
    Computes product-form Cholesky factorization of a matrix I +
    V*diag(a)*V.T of order n and with V n-by-m, i.e,

      I + V*diag(a)*V.T = L_1*L_2*...*L_k*L_k'*...*L_2'*L_1'.

    The arrays L and B must each be of length (at least) n*k.
   */
  int i,j;
  double tmp, alpha;
  double *v,*l,*b;
  char cN='N';
  
  for (i=0;i<*k;i++) {
    
    v = V+i*(*ldv); 
    b = B+i*(*ldb); 
    l = L+i*(*ldl);
    alpha = -a[i];
    
    for (j=0;j<*n;j++) {
      tmp = 1.0-alpha*v[j]*v[j];
      if (tmp<=0) return -1;
      l[j] = sqrt(tmp);
      b[j] = -alpha*v[j]/l[j];
      alpha += b[j]*b[j];
    }
    
    for (j=i+1;j<*k;j++) {
      ddrsv(n,&cN,v,l,b,V+j*(*ldv));
    }

  }
  return 0;
}

void dpfsv(const int *n, const int *k, const char *trans, double * restrict V, const int *ldv, double * restrict L, const int *ldl, double * restrict B, const int *ldb, double * restrict x) {  
  /*
    Solves a system 

        L_1*L_2*...*L_k*y = x      if trans == 'N'

    or 

        L_k'*...*L_2'*L_1'*y = x   if trans == 'T'
 
    where L_1*L_2*...*L_k*L_k'*...*L_2'*L_1' is the product-form
    Cholesky-factorization of a matrix I + V*diag(a)*V.T of order
    n. On exit, x contains the solution y.
   */

  int i;  
  if (*trans == 'N') {
    for (i=0;i<*k;i++) ddrsv(n,trans,V+i*(*ldv),L+i*(*ldl),B+i*(*ldb),x);
  }
  else if (*trans == 'T') {
    for (i=*k-1;i>=0;i--) ddrsv(n,trans,V+i*(*ldv),L+i*(*ldl),B+i*(*ldb),x);
  }
  return;
}
		
void dpfmv(const int *n, const int *k, const char *trans, double * restrict V, const int *ldv, double * restrict L, const int *ldl, double * restrict B, const int *ldb, double * restrict x) {  
  /*
    Computes the matrix-vector product

        x := L_1*L_2*...*L_k*x      if trans == 'N'

    or 

        x := L_k'*...*L_2'*L_1'*x   if trans == 'T'
 
    where L_1*L_2*...*L_k*L_k'*...*L_2'*L_1' is the product-form
    Cholesky-factorization of a matrix I + V*diag(a)*V.T of order
    n.
   */
  int i;  
  if (*trans == 'N') {
    for (i=*k-1;i>=0;i--) ddrmv(n,trans,V+i*(*ldv),L+i*(*ldl),B+i*(*ldb),x);
  }
  else if (*trans == 'T') {
    for (i=0;i<*k;i++) ddrmv(n,trans,V+i*(*ldv),L+i*(*ldl),B+i*(*ldb),x);
  }
  return;
}

File: src/python/pfcholesky.py
import chompack as cp
from chompack.symbolic import cspmatrix, symbolic
from cvxopt import matrix, spmatrix, lapack

try:
    from chompack.cbase import pfchol, pftrmm, pftrsm
except:
    from math import sqrt
    def _ddrsv(n,trans,v,l,b,x):
        s = 0.0
        if trans=='N':
            for k in range(n):
                x[k] = (x[k] - v[k]*s)/l[k]
                s += b[k]*x[k]
        elif trans=='T':
            for k in range(n-1,-1,-1):
                x[k] = (x[k]-b[k]*s)/l[k]
                s += x[k]*v[k]
        else: raise ValueError("trans must be 'N' or 'T'")
        return

    def _ddrmv(n,trans,v,l,b,x):
        s = 0.0
        if trans=='N':
            for k in range(n):
                tmp = x[k]*b[k]
                x[k] = l[k]*x[k] + v[k]*s
                s += tmp
        elif trans=='T':
            for k in range(n-1,-1,-1):
                tmp = x[k]*v[k]
                x[k] = l[k]*x[k]+b[k]*s
                s += tmp
        else: raise ValueError("trans must be 'N' or 'T'")
        return

    def _dpftrf(n,k,a,V,L,B):
        for i in range(k):
            alpha = -a[i];
            v = V[:,i]
            l = L[:,i]
            b = B[:,i]
            
            for j in range(n):
                tmp = 1.0 - alpha*v[j]*v[j]
                if (tmp <= 0): return -1
                l[j] = sqrt(tmp)
                b[j] = -alpha*v[j]/l[j]
                alpha += b[j]*b[j]
            L[:,i] = l
            B[:,i] = b
                
            for j in range(i+1,k):
                vj = V[:,j]
                _ddrsv(n,'N',v,l,b,vj)
                V[:,j] = vj
        return
    
    def _dpfsv(n,k,trans,V,L,B,x):
        if trans=='N':
            for i in range(k):
                v = V[:,i]
                l = L[:,i]
                b = B[:,i]
                _ddrsv(n,trans,v,l,b,x);
                V[:,i] = v
        elif trans=='T':            
            for i in range(k-1,-1,-1):
                v = V[:,i]
                l = L[:,i]
                b = B[:,i]
                _ddrsv(n,trans,v,l,b,x)
                V[:,i] = v
        else: raise ValueError("trans must be 'N' or 'T'")
        return
        
    def _dpfmv(n,k,trans,V,L,B,x):
        if trans=='N':
            for i in range(k-1,-1,-1):
                v = V[:,i]
                l = L[:,i]
                b = B[:,i]
                _ddrmv(n,trans,v,l,b,x);
                V[:,i] = v
        elif trans=='T':            
            for i in range(k):
                v = V[:,i]
                l = L[:,i]
                b = B[:,i]
                _ddrmv(n,trans,v,l,b,x)
                V[:,i] = v
        else: raise ValueError("trans must be 'N' or 'T'")
        return
    
    def pfchol(alpha,V,L,B):
        _dpftrf(V.size[0],V.size[1],alpha,V,L,B)
        return

    def pftrsm(V,L,B,U,trans='N'):
        for i in range(U.size[1]):
            u = U[:,i]
            _dpfsv(U.size[0],L.size[1],trans,V,L,B,u)
            U[:,i] = u
        return

    def pftrmm(V,L,B,U,trans='N'):
        for i in range(U.size[1]):
            u = U[:,i]
            _dpfmv(U.size[0],L.size[1],trans,V,L,B,u)
            U[:,i] = u
        return
    
class pfcholesky(object):
    """
    Supernodal multifrontal product-form Cholesky factorization:

    .. math::
    
        X + V \mathrm{diag}(a) V^T = L_m \cdots L_1L_0 L_0^TL_1^T \cdots L_m^T

    where :math:`X = L_0L_0^T` is of order n and :math:`V` is n-by-m.

    :param X:       :py:class:`cspmatrix` or :py:class:`spmatrix`
    :param V:       n-by-m matrix
    :param a:       m-by-1 matrix (optional, default is vector of ones)
    :param p:       n-by-1 matrix (optional, default is natural ordering)
    """

    def __init__(self,X,V,a=None,p=None):

        self._n = X.size[0]
        self._V = +V        
        assert X.size[0] == X.size[1], 'X must be a square matrix'
        assert X.size[0] == V.size[0], 'V must have have the same number of rows as X'
        assert isinstance(V, matrix), 'V must be a dense matrix'
        
        if isinstance(X, cspmatrix) and X.is_factor is False:
            self._L0 = X.copy()            
            cp.cholesky(self._L0)
            cp.trsm(self._L0, self._V)
        elif isinstance(X, cspmatrix) and X.is_factor is True:
            self._L0 = X
            cp.trsm(self._L0, self._V)
        elif isinstance(X, spmatrix):
            symb = symbolic(X, p = p)
            self._L0 = cspmatrix(symb) + X
            cp.cholesky(self._L0)
            cp.trsm(self._L0, self._V)
        elif isinstance(X, matrix):
            raise NotImplementedError
        else:
            raise TypeError

        if a is None: a = matrix(1.0,(self._n,1))
        self._L = matrix(0.0,V.size)
        self._B = matrix(0.0,V.size)
        pfchol(a,self._V,self._L,self._B)
        return

    def __repr__(self):
        return "<%ix%i product-form Cholesky factor, r=%i, tc='%s'>"\
              % (self._L0.size[0],self._L0.size[0],self._V.size[1],self._L0.blkval.typecode) 

    def trsm(self,B,trans='N'):
        r"""
        Solves a triangular system of equations with multiple righthand
        sides. Computes

        .. math::

            B &:= L^{-1} B  \text{ if trans is 'N'}

           B &:= L^{-T} B  \text{ if trans is 'T'} 
        """
        
        if trans=='N':
            cp.trsm(self._L0,B)
            pftrsm(self._V,self._L,self._B,B,trans='N')
        elif trans=='T':
            pftrsm(self._V,self._L,self._B,B,trans='T')
            cp.trsm(self._L0,B,trans='T')
        elif type(trans) is str:
            raise ValueError("trans must be 'N' or 'T'")
        else:
            raise TypeError("trans must be 'N' or 'T'")
        return

    def trmm(self,B,trans='N'):
        r"""
        Multiplication with product-form Cholesky factor. Computes
    
        .. math::
    
            B &:= L B    \text{ if trans is 'N'}

            B &:= L^T B  \text{ if trans is 'T'}
        """
        
        if trans=='N':
            pftrmm(self._V,self._L,self._B,B,trans='N')
            cp.trmm(self._L0,B)            
        elif trans=='T':
            cp.trmm(self._L0,B,trans='T')
            pftrmm(self._V,self._L,self._B,B,trans='T')
        elif type(trans) is str:
            raise ValueError("trans must be 'N' or 'T'")
        else:
            raise TypeError("trans must be 'N' or 'T'")
        return

        
